### PR TITLE
fix(cascade): preserve empty attention states during merges

### LINKED ISSUE #4051: [Bug]: Cascade state merging returns NaN when all input states are empty
## Observed behavior

FlashInfer represents an empty attention state as zero V with logsumexp equal to negative infinity. Merging two such states currently produces NaN for both the attention output and logsumexp.

This affects `merge_state`, `merge_state_in_place`, and reductions using `state_t::merge`, including `merge_states` and persistent attention reductions.

## Reproduction

The official trace reference reproduces the failure:

```python
import torch
from flashinfer.trace.templates.cascade import merge_state_trace

v_a = torch.zeros((1, 1, 64), dtype=torch.float16)
v_b = torch.zeros_like(v_a)
s_a = torch.full((1, 1), -torch.inf, dtype=torch.float32)
s_b = torch.full_like(s_a, -torch.inf)

v, s = merge_state_trace.reference(v_a, s_a, v_b, s_b)

print(torch.isnan(v).any().item())  # True
print(torch.isnan(s).any().item())  # True
```

The merge formula evaluates:

```text
s_a = s_b = -inf
s_max = max(s_a, s_b) = -inf
exp2(s_a - s_max) = exp2(-inf - -inf) = NaN
```

## Expected behavior

The union of empty attention states should remain empty:

```text
merge((0, -inf), (0, -inf)) = (0, -inf)
```

This matches the existing behavior for zero input index sets.

## Impact and scope

An empty or fully masked partition can contaminate subsequent cascade reductions with NaN. Finite-state and single-empty merges are unaffected.

## Success criteria

- `merge_state` and `merge_state_in_place` preserve the empty state.
- Both `merge_states` launch strategies preserve the empty state.
- `state_t::merge` does not evaluate `-inf - -inf`.
- Existing finite-state behavior remains unchanged.

### PR DESCRIPTION
**Pull request by morluto**
## 📌 Description

An empty attention state is represented by zero V and an LSE of `-inf`. Merging two empty states previously evaluated:

```text
s_a = s_b = -inf
s_max = max(s_a, s_b) = -inf
exp2(s_a - s_max) = exp2(-inf - -inf) = NaN
```

The required identity is:

```text
merge((0, -inf), (0, -inf)) = (0, -inf)
```

This change:

- returns zero V and `-inf` LSE from the binary merge kernels when both inputs are empty;
- makes a double-empty `state_t::merge` a neutral no-op, covering `merge_states` and other persistent reductions using the shared state;
- preserves the same neutral state in Triton binary, in-place, multi-state, and variable-length merges;
- updates the cascade trace references and checked-in trace definitions;
- adds regressions through the CUDA and Triton cascade merge APIs.

Finite-state and single-empty behavior are unchanged.

## 🔍 Related Issues

Fixes #4051.

## 🧪 Tests

- `python3 -m pytest -q tests/trace/test_merge_state_reference_correctness.py -k preserves_empty_state` — 3 passed
- `python3 -m pytest -q tests/trace/test_rendered_source_standalone.py` — 2 passed
- Added CUDA and Triton empty-state regressions for `merge_state`, `merge_state_in_place`, `merge_states`, and Triton variable-length merging
- Focused pre-commit hooks and `git diff --check` passed

## Reviewer Notes

The central invariant is in `state_t::merge`: two states with maximum logit `-inf` contain no attention mass, so their merge must preserve the initialized zero/`-inf` state.

The binary kernels do not use `state_t`, so they implement the same result explicitly.

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed cascade attention state merging for “all-empty” inputs, where log-sum-exp maxima are `-inf`.
  * Merged outputs now remain numerically valid: value vectors stay zeros and corresponding scores remain `-inf`, avoiding NaNs.
  * Applied consistently across reference, CUDA, and Triton merge variants (including in-place and multi-state merging).

* **Tests**
  * Added new correctness checks to ensure empty-state semantics are preserved for regular, in-place, and multi-state merges.
  * Updated trace/reference implementations to cover the “both empty” edge case.

### CODE REVIEW

File: tests/utils/test_triton_cascade.py [hunk 1]
```diff
@@ -6,6 +6,75 @@
 from tests.test_helpers.test_helpers import skip_on_gpu_arch_error
 
 
+@skip_on_gpu_arch_error
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (morluto) [reply tests/utils/test_triton_cascade.py:10]: This decorator is already present immediately above the parametrization (`@skip_on_gpu_arch_error` on line 9, followed by `@pytest.mark.parametrize` on line 10), so non-supported architectures use the same skip path as the other tests in this file. No change is needed.
